torchtune/utils/_activations_offloading/offload_ctx_mgr.py

```diff
@@ -0,0 +1,278 @@
+                        device=torch.device("cpu"),
+                    )
+
+                    cpu_tensor.copy_(activation, non_blocking=True)
+                    self.tracker[tensor_id] = (
+                        cpu_tensor,
+                        True,
+                    )  # True = (in future) modified
+
+                if use_streams:
+                    event = self.s1.record_event()
+
+                    # Stash to keep activation alive til s1 is done
+                    self.fwd_stash[tensor_id] = (activation, event)
+            else:
+                self.tracker[tensor_id] = (
+                    activation,
+                    False,
+                )  # False = not modified, tensor is as is
+
+            return tensor_id
+
+        def unpack_tensor(unpack_tensor_id: int) -> torch.Tensor:
+            # backward pass - we are called with the tensor_id, which
+            # we will use to retrieve the saved/offloaded tensor
+            if self.is_first_backward_call:
+                if use_streams:
+
+                    def wait_and_del_remaining_references() -> None:
+                        for id in [k for k in self.bwd_tensor_stash.keys()]:
+                            event = self.bwd_ev_stash[id]
+                            self.s1.wait_event(event)
+                            del self.bwd_tensor_stash[id]
+
+                    # Register a callback to the end of autograd to clean everything up
+                    torch.autograd.variable.Variable._execution_engine.queue_callback(
+                        wait_and_del_remaining_references
+                    )
+
+                if self.is_first_forward_pass:
+                    self.is_first_forward_pass = False
+                    if self.use_pin_memory:
+                        verify_sufficient_virtual_memory()
+
+                self.is_first_backward_call = False
+                self.is_first_forward_call = True
+
+            assert (
+                unpack_tensor_id in self.tracker
+            ), f"untracked tensor with id {unpack_tensor_id}"
+
+            maybe_gpu_tensor, modified = self.tracker[unpack_tensor_id]
+            if modified:
+                if not use_streams:
+                    gpu_tensor = maybe_gpu_tensor.to(device="cuda", non_blocking=True)
+                    maybe_gpu_tensor = gpu_tensor
+                else:
+                    # Update data on the current autograd node
+                    graph_id = torch._C._current_graph_task_id()
+                    node = torch._C._current_autograd_node()
+                    prev_node_ids = []
+
+                    # Init graph id if it was uninitialized
+                    if not self.curr_graph_id:
+                        self.curr_graph_id = graph_id
+
+                    # If we're on a new node, mark prev node's tensors to be freed later
+                    if (
+                        graph_id == self.curr_graph_id
```

Comment:
Maybe this is just cause I don't understand what `_current_graph_task_id` represents, but when would equality not hold here?

Reply:
Different backwards could maybe be occurring at the same time--this makes sure that we're tracking for the right graph. I don't know many details on this and followed @albanD's advice here.